### PR TITLE
[WIP] Quantile Regression

### PR DESCRIPTION
**Pull request by david-waterworth**
I hereby agree to the terms of the CLA available at: [link](https://yandex.ru/legal/cla/?lang=en)

This is a work in progress to improve the performance of Quantile Regression. It removes the restriction of only performing a single leaf estimation iteration for Quantile Loss (and other).

The issue is the quantile loss function has constant (per sample) leaf first derivatives and zero 2nd derivatives. So gradient convergence can be slow (can be improved by scaling and choosing a better starting point). Both sklearn and lightgbm approach this issue by bypassing the leaf estimation and replacing with the exact value (since the value which minimises the quantile loss function is the alpha-quantile).

### CODE REVIEW

**Comment by david-waterworth** [comment]
Also an open question I have is: is it better to solve the leaf estimates exactly at each iteration. Or should we do a line search and find the best approximation in the original direction and then move on to the next iteration.

i.e. at the moment we do a backward line search. But since the objective is (hyper) V shaped, if the initial step size is to large we jump from one side of the V to the other. The Armijo condition is usually satisfied so we move to the next leaf search direction. If instead we search for the point were the objective is minimised in the original search direction then stop, at least the leaf estimates are consistent with the gradients used in the splitting step even if they don't fully minimise the objective?

**Comment by david-waterworth** [comment]
And some motivation:

```
import numpy as np
from catboost import CatBoostRegressor
from catboost import Pool

X = np.array([1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2], dtype=np.float32)
y = np.array([1,2,3,4,5,6,7,8,9,10,21,22,23,24,25,26,27,28,29,30], dtype=np.float32)
alpha=0.75

train_data = Pool(X, y)
a0 = np.mean(y)
train_data.set_baseline(np.repeat(a0,len(y)))

model = CatBoostRegressor(task_type='GPU',l2_leaf_reg=0.5,leaf_estimation_backtracking='Armijo',leaf_estimation_iterations=10,learning_rate=1,iterations=1,loss_function=f'Quantile:alpha=0.75',boosting_type='Plain',logging_level='Verbose')
model.fit(train_data);

model.predict(X)+a0
```

catboost 0.13.0 takes about 50 iterations / 37ms to converge (and it's not quite reached the minima)

> 0:	learn: 4.7023811	total: 1.2ms	remaining: 59ms
> 1:	learn: 4.4047623	total: 2ms	    remaining: 47.9ms
> 
> 44:	learn: 0.9321432	total: 33.7ms	remaining: 3.75ms
> 45:	learn: 0.9297623	total: 34.3ms	remaining: 2.99ms
> 46:	learn: 0.9285709	total: 35.2ms	remaining: 2.25ms
> 47:	learn: 0.9273813	total: 36ms	    remaining: 1.5ms
> 48:	learn: 0.9261900	total: 36.6ms	remaining: 747us
> 49:	learn: 0.9261909	total: 37.2ms	remaining: 0us
> array([ 7.9761904,  7.9761904,  7.9761904,  7.9761904,  7.9761904,
>         7.9761904,  7.9761904,  7.9761904,  7.9761904,  7.9761904,
>        27.9761908, 27.9761908, 27.9761908, 27.9761908, 27.9761908,
>        27.9761908, 27.9761908, 27.9761908, 27.9761908, 27.9761908])

Doing 25 gradient steps using armijo backtracking (and hacking the 2nd derivative to ensure it's 1), requires 1 iteration and 2ms - so ~20x faster. 

> 0:	learn: 0.9250000	total: 1.83ms	remaining: 0us
> array([ 8.        ,  8.        ,  8.        ,  8.        ,  8.        ,
>         8.        ,  8.        ,  8.        ,  8.        ,  8.        ,
>        27.99999905, 27.99999905, 27.99999905, 27.99999905, 27.99999905,
>        27.99999905, 27.99999905, 27.99999905, 27.99999905, 27.99999905])

OR

Doing 1 gradient step, but with the fixed 2nd derivative), requires ~10 iterations and 7.5ms so around 5x improvement.

0:	learn: 2.9166664	total: 1.14ms	remaining: 10.3ms
1:	learn: 1.4583332	total: 1.78ms	remaining: 7.14ms
2:	learn: 1.1083332	total: 2.36ms	remaining: 5.5ms
3:	learn: 0.9583333	total: 2.92ms	remaining: 4.38ms
4:	learn: 0.9416666	total: 3.61ms	remaining: 3.61ms
5:	learn: 0.9333334	total: 4.16ms	remaining: 2.77ms
6:	learn: 0.9333333	total: 4.76ms	remaining: 2.04ms
7:	learn: 0.9333334	total: 5.45ms	remaining: 1.36ms
8:	learn: 0.9333333	total: 6.26ms	remaining: 695us
9:	learn: 0.9333334	total: 7.43ms	remaining: 0us
array([ 8.16666681,  8.16666681,  8.16666681,  8.16666681,  8.16666681,
        8.16666681,  8.16666681,  8.16666681,  8.16666681,  8.16666681,
       28.16666666, 28.16666666, 28.16666666, 28.16666666, 28.16666666,
       28.16666666, 28.16666666, 28.16666666, 28.16666666, 28.16666666])

**Comment by david-waterworth** [comment]
I think it's worth merging this as is, it improves convergence quite a bit. I don't think there' much more I can do without quite a bit of additional work to provide an `exact` leaf solution mode like catboost (i.e. for loss functions which have a known closed form, use that in place of newton / gradient). Adding this is more work than I have time for at this point in time.

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request created: [801735](https://a.yandex-team.ru/review/801735)

**Comment by arcadia-devtools** [comment]
Check [801735](https://a.yandex-team.ru/review/801735) failed: 1 test(s) broken

**Comment by annaveronika** [comment]
Sorry for a late answer! Thanks a lot for noticing that this restriction is not needed. The pr will be merged soon (failing test is a flaky one). It would be also nice to improve the mode further, if you will have time for that.

**Comment by david-waterworth** [comment]
Thanks @annaveronika, when I get some more time I do want to take another look. I'm thinking perhaps add 2 methods to the loss function class

bool HasArgMin
float ArgMin(float[] targets, float[] estimates)

So If HasArgMin==true, then instead of invoking the leaf estimation gradient solver, call ArgMin which returns the argument which minimises the loss function (i.e. in this case the alpha quantile). I think you'd probably still use the gradient method to build the tree.
